sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/joins.scala

```diff
@@ -204,6 +204,26 @@ object EliminateOuterJoin extends Rule[LogicalPlan] with PredicateHelper {
 
   def apply(plan: LogicalPlan): LogicalPlan = plan.transformWithPruning(
     _.containsPattern(OUTER_JOIN), ruleId) {
+    case p @ Project(projectList, Filter(cond, j @ Join(left, right, LeftOuter, _, _)))
+        if projectList.forall(canEvaluate(_, left)) =>
+      val notNullAttrs =
+        AttributeSet(right.constraints.collect { case IsNotNull(a: Attribute) => a })
+      if (notNullAttrs.nonEmpty) {
+        val (isNullConditions, rest) = splitConjunctivePredicates(cond).partition {
+          case IsNull(a: Attribute) => notNullAttrs.contains(a)
+          case _ => false
+        }
+        if (isNullConditions.nonEmpty &&
+          rest.forall(_.references.intersect(right.outputSet).isEmpty)) {
```

Comment:
Please add more test cases to cover the different code paths.

Reply:
Thanks @wangyum , I have updated the UT.

Another question, for this UT,  could we optimize `testRelation1.subquery("y").where($"d".isNotNull).where($"d".isNull && $"f" > 10)` to `EmptyRelation` to skip data scan?